content/param_id_func/contents.lr

```diff
@@ -0,0 +1,470 @@
+slightly simplified explanation, but that's all we need to know for now.
+
+Here's a code example for clarification:
+
+```python
+@pytest.mark.parametrize(
+    'number, word', [
+        (1, '1'),
+        (3, 'fizz'),
+        (5, 'buzz'),
+        (8, '8'),
+        (10, 'buzz'),
+        (15, 'fizzbuzz'),
+    ]
+)
+def test_fizzbuzz(number, word):
+    assert fizzbuzz(number) == word
+```
+
+There is a single test function in above code example. It calls a function
+called ``fizzbuzz`` with an integer and checks that its return value matches a
+certain string.
+
+Now when we run this, **pytest** creates a number of test items under the hood.
+
+- Item with function: ``test_fizzbuzz``, number: ``1``, word: ``'1'``
+- Item with function: ``test_fizzbuzz``, number: ``3``, word: ``'fizz'``
+- Item with function: ``test_fizzbuzz``, number: ``5``, word: ``'buzz'``
+- Item with function: ``test_fizzbuzz``, number: ``8``, word: ``'8'``
+- Item with function: ``test_fizzbuzz``, number: ``10``, word: ``'buzz'``
+- Item with function: ``test_fizzbuzz``, number: ``15``, word: ``'fizzbuzz'``
+
+This is exactly what we can see in the test result log when running **pytest**
+in normal mode (not quiet or verbose).
+
+# what are ids?
+
+Now **pytest** lets us add *ids* to test items to make them easier to
+distinguish in the log, especially in verbose mode.
+
+## auto-generated ids
+
+If you leave it up to **pytest** to generate ids, you will see that most of
+them are pretty good, but others feel somewhat random. The reason being that
+*non-primitive* types such as ``dict``, ``list``, ``tuple`` or instances of
+``classes`` are non trivial to translate into ids.
+
+The following test is parametrized and the id for a test item will be a
+representation for **each individual parameter** - joined with a ``-``.
+
+```python
+class CookiecutterTemplate:
+    def __init__(self, name, url):
+        self.name = name
+        self.url = url
+
+PYTEST_PLUGIN = CookiecutterTemplate(
+    'pytest-plugin',
+    'https://github.com/pytest-dev/cookiecutter-pytest-plugin',
+)
+
+@pytest.mark.parametrize(
+    'a, b',
+    [
+        (1, {'Two Scoops of Django': '1.8'}),
+        (True, 'Into the Brambles'),
+        ('Jason likes cookies', [1, 2, 3]),
+        (PYTEST_PLUGIN, 'plugin_template'),
+    ]
+)
+def test_foobar(a, b):
+    assert True
+```
+
+``$ pytest -v`` produces the following report:
+
+```text
+============================ test session starts =============================
+collecting ... collected 4 items
+
+test_ids.py::test_foobar[1-b0] PASSED
+test_ids.py::test_foobar[True-Into the Brambles] PASSED
+test_ids.py::test_foobar[Jason likes cookies-b2] PASSED
+test_ids.py::test_foobar[a3-plugin_template] PASSED
+
+========================== 4 passed in 0.03 seconds ==========================
+```
+
+As you can see whenever **pytest** encounters one of the *non-primitives* it
+uses the fixture name with a suffix instead.
```

Comment:
Hmm fair point 😁

TBH it's the same to me. But I can see how this could be confusing. I'll change that!